### PR TITLE
rewrite to lexer/parser approach

### PR DESCRIPTION
**Pull request by hairmare**
This rewrites most of the parser to a doctrine/lexer based approach.

This should be able to pave the way for #6 as well as contribute to GRV-1289.

Implements a heap of rql operations in their basic non sugary form.
- [x] `eq()`, `ne()`, `lt()`, `gt()`, `lte()` and `gte()` support
- [x] `and()` and `or()` support
- [x] implement `like()` support
- [x] fix failing `or()` test
- [x] implement `in()`, `out()`
- [x] implement `limit()`

These are things we also need to do. I'd rather do them in different PRs so we can get 2.0.0-alpha1 out of the door soonish.
- refactor AST -> we need `AST\EqOperation` based on `AST\AbstractPropertyOperation` and so on. This also adds some more interfaces. The types we need are probably propertyOp, queryOp, visitOp and sortOp, though the last two need checking with other possible methods we'll be needing. (would be better suited for a separate PR)
- implement syntax sugar variants, probably needs some parser refactoring
- re-integrate with graviton by way of a wrapper-bundle (ie. like serialiser does, also tracked as GRV-1283, [http://gnugat.github.io/2014/10/29/sf2-bundle-standalone.html]) 
- fork mongo visitor into libgraviton/php-rql-parser-visitor-mongo-odm so everyone is treated the same

This also needs GRV-1283 to be useable with graviton. Along with GRV-1283 it fixes GRV-1289.

### CODE REVIEW

File: src/Graviton/Rql/Visitor/MongoOdm.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+     * @var string[]
+     */
+    private $operationMap = array(
+        'eq' => 'equals',
+        'ne' => 'notEqual',
+        'lt' => 'lt',
+        'gt' => 'gt',
+        'lte' => 'lte',
+        'gte' => 'gte'
+);
```
Comment (narcoticfresh) [inline src/Graviton/Rql/Visitor/MongoOdm.php]: The keys here should reference to the corresponding Lexer Constants..

File: src/Graviton/Rql/Visitor/MongoOdm.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    /**
+     * @var string<string>
+     */
+    private $internalMethods = array(
+        'sort' => 'visitSort',
+        'like' => 'visitLike',
+        'limit' => 'visitLimit',
+    );
```
Comment (narcoticfresh) [inline src/Graviton/Rql/Visitor/MongoOdm.php:44]: Same here, keys should refer to some constants..

File: src/Graviton/Rql/Visitor/MongoOdm.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        $expr = $this
+            ->queryBuilder
+            ->expr()
+            ->field($operation->getProperty());
+
+        if ($operation->getName() == 'eq') {
+            $expr = $expr->equals($operation->getValue());
+        } elseif ($operation->getName() == 'ne') {
+            $expr = $expr->notEqual($operation->getValue());
+        }
```
Comment (narcoticfresh) [inline src/Graviton/Rql/Visitor/MongoOdm.php:124]: 'eq' and 'ne' and so on should also refer to constants..

File: src/Graviton/Rql/Visitor/MongoOdm.php (hunk 1 above)
Comment (hairmare) [inline src/Graviton/Rql/Visitor/MongoOdm.php]: These do not use the Lexer::T_\* constants as not to couple the visitor to the Lexer. I think the visitor should only depend on the AST. The "magic" strings here come from the ASTs `getType()` operation.

I should probably refactor the AST so it can express this as a composition of various classes. That would then lead to `AST\Operation\Eq` and so on. I'm not convinced that refactoring the AST will make the code more reusable though.

File: src/Graviton/Rql/Visitor/MongoOdm.php (hunk 2 above)
Comment (hairmare) [inline src/Graviton/Rql/Visitor/MongoOdm.php:44]: Same as above, the Visitor should not be coupled to the Lexer. This is based on the AST.
I'll be refactoring the AST as part of a different PR.

File: src/Graviton/Rql/Visitor/MongoOdm.php (hunk 3 above)
Comment (hairmare) [inline src/Graviton/Rql/Visitor/MongoOdm.php:124]: Again, this should not couple the visitor to the lexer. This method does need work to be usable though.

**Comment by narcoticfresh** [comment]
What about defining constants in the interface which the Visitor can refer to?
Then we had a defined set of operations and types the Visitors can use and it's not coupled to the Lexer.. even if it's duplication it allows for a better style in the Visitor..

**Comment by hairmare** [comment]
I don't know about using `const` in an `interface`. I was under the impression that that would be invalid.

What do you think about refactoring the AST so the info you're looking at is explicitly added to the graph?

**Comment by hairmare** [comment]
I just removed the [WIP] stamp (we should switch to WIP labels btw).

This PR is getting larger as it should already ;)

To summarize:
- The is practically a from the grounds up rewrite. I did keep small parts though.
- Its based on doctrine/lexer that has been stable for a loooong time now and is heavily depended on by most doctrine projects (by way of the annotation-parser btw).
- This needs a refactor of the internal AST asap.
- more stuff thats mentioned in the PRs original text (and either covered by GRV or things I'll create issues on)

Also, this isn't strictly at a usable state right now. We need to write an additional bundle for integrating this with symfony. This only has doctrine/lexer and the querybuilder from doctrine/mongo-odm as runtime dependency. The internal AST refactor does not interact with this.

Basically, I want to merge it soonish so we can release these results as `2.0.0-alpha1` and get on with building a properly integrated 2.0.0 release.

@narcoticfresh I started an AST refactor over here: 66e3fb7c10b6a70551a50273eda73e1b766070bf
We should talk the AST structure through, before I go all in.

**Comment by narcoticfresh** [comment]
Well, we're mixing too much at once ;-)
First off, a Symfony integration obviously needs to be a separate project with a separate composer.json and all.. not scope of this project here..

Regarding the AST refactoring: I don't think we should blow this up too much. Wouldn't it be more leaner to just define an abstract class somewhere that just contains constants to which both parts can refer?

**Comment by hairmare** [comment]
The symfony bundle won't be part of this PR. I'm just mentioning it since we need it so php-rql-parser is only with graviton again.

I've identified some grouping that make sense in the AST (making for easier visitors that don't use magic strings). We can discuss those in a new PR as soon as this gets merged.
